linalg/constraints.hpp

```diff
@@ -0,0 +1,462 @@
+   mutable Vector multiplier_sol;
+   mutable Vector workb;
+   mutable Vector workx;
+
+private:
+   void Initialize();
+};
+
+
+/** @brief Perform elimination of a single constraint.
+
+    See EliminationProjection, EliminationCGSolver
+
+    This keeps track of primary / secondary tdofs and does small dense block
+    solves to eliminate constraints from a global system.
+
+    \f$ B_s^{-1} \f$ maps the lagrange space into secondary dofs, while
+    \f$ -B_s^{-1} B_p \f$ maps primary dofs to secondary dofs. */
+class Eliminator
+{
+public:
+   Eliminator(const SparseMatrix& B, const Array<int>& lagrange_dofs,
+              const Array<int>& primary_tdofs,
+              const Array<int>& secondary_tdofs);
+
+   const Array<int>& LagrangeDofs() const { return lagrange_tdofs; }
+   const Array<int>& PrimaryDofs() const { return primary_tdofs; }
+   const Array<int>& SecondaryDofs() const { return secondary_tdofs; }
+
+   /// Given primary displacements, return secondary displacements
+   /// This applies \f$ -B_s^{-1} B_p \f$.
+   void Eliminate(const Vector& in, Vector& out) const;
+
+   /// Transpose of Eliminate(), applies \f$ -B_p^T B_s^{-T} \f$
+   void EliminateTranspose(const Vector& in, Vector& out) const;
+
+   /// Maps Lagrange multipliers to secondary displacements,
+   /// applies \f$ B_s^{-1} \f$
+   void LagrangeSecondary(const Vector& in, Vector& out) const;
+
+   /// Transpose of LagrangeSecondary()
+   void LagrangeSecondaryTranspose(const Vector& in, Vector& out) const;
+
+   /// Return \f$ -B_s^{-1} B_p \f$ explicitly assembled in mat
+   void ExplicitAssembly(DenseMatrix& mat) const;
+
+private:
+   Array<int> lagrange_tdofs;
+   Array<int> primary_tdofs;
+   Array<int> secondary_tdofs;
+
+   DenseMatrix Bp;
+   DenseMatrix Bs;  // gets inverted in place
+   LUFactors Bsinverse;
+   DenseMatrix BsT;   // gets inverted in place
+   LUFactors BsTinverse;
+   Array<int> ipiv;
+   Array<int> ipivT;
+};
+
+
+/** Collects action of several Eliminator objects to perform elimination of
+    constraints.
+
+    Works in parallel, but each Eliminator must be processor local, and must
+    operate on disjoint degrees of freedom (ie, the primary and secondary dofs
+    for one Eliminator must have no overlap with any dofs from a different
+    Eliminator). */
+class EliminationProjection : public Operator
+{
+public:
+   EliminationProjection(const Operator& A, Array<Eliminator*>& eliminators);
+
+   void Mult(const Vector& x, Vector& y) const override;
+
+   void MultTranspose(const Vector& x, Vector& y) const override;
+
+   /** @brief Assemble this projector as a (processor-local) SparseMatrix.
+
+       Some day we may also want to try approximate variants. */
+   SparseMatrix * AssembleExact() const;
+
+   /** Given Lagrange multiplier right-hand-side \f$ g \f$, return
+       \f$ \tilde{g} \f$ */
+   void BuildGTilde(const Vector& g, Vector& gtilde) const;
+
+   /** After a solve, recover the Lagrange multiplier. */
```

Comment:
It appears you still refer to a few variables here as displacement variables rather than the abstract dofs that Jamie had mentioned earlier in a comment.